# Brown CBC's Metagenomics Workshop Materials

[![Docs](https://img.shields.io/badge/docs-stable-blue.svg?style=flat-square)](https://compbiocore.github.io/metagenomics-workshop)
[![nbviewer](https://img.shields.io/badge/jupyter_notebooks-nbviewer-purple.svg?style=flat-square)](http://nbviewer.jupyter.org/github/compbiocore/metagenomics-workshop/tree/master/docs/src/notebooks)
[![License](https://img.shields.io/aur/license/yaourt.svg)](https://raw.githubusercontent.com/compbiocore/metagenomics-workshop/master/LICENSE)


## Goals

Session 1 Fundamental concepts

This will include the overall goal of metagenomics, the types of metagenomics experiments, and considerations for experimental design.

Session 2 Workflows and tools for analysis

This will include workflows and tools for analysis as well as some common pitfalls and how to troubleshoot them.

## Workshop Topic Overview

Session 1 Analysis of 16S data using QIIME - Kellyanne Duncan

This will be a short tutorial on the analysis of 16S data using QIIME with an overview of quality filtering and examples of some different aspects of analyses.

Session 2 Analysis of 16S data using DADA2 - Damien Cabral

This will be a short tutorial on the analysis of 16S data using DADA2 that will include DADA2 background and core principles, DADA 2 workflow, and a brief introduction to analysis in Phyloseq.

### **[Notebook](http://nbviewer.jupyter.org/github/compbiocore/metagenomics-workshop/tree/master/docs/src/notebooks)**

To access a read-only version of the workshop notebooks (complete with full output), please click the above link; you will be redirected to Jupyter's NBViewer utility.  Simply click the name of the notebook and it will be rendered as HTML.

### Authors

Kellyanne Duncan, Damien Cabral

### Contact

For assistance, contact cbc-help@brown.edu - this is our general help line, so please specify that your issue is with this site's contents

### Original Time and Location

> Date: October 2017

> Location: Campus Center
